### PR TITLE
Implemented gRPC binding for `getDataAtPar` method

### LINKED ISSUE #3538: Implement replace API for listenForDataAtName
## Overview

**Current shortcoming of `listenForDataAtName`**

When users try to request `listenForDataAtName` API, the node would traverse blocks based on the request param `depth` which is very low efficiency and waste.  Because it requires traversal through blocks _depth_ is limited to only 3 blocks by default to prevent attacking the node in this way.

We want to remove this API and provide improved way for users to make request for data at some name (channel). This also includes data sent on `rho:rchain:deployId`.

### Impact

Any clients which use this API should consider move to the new API.

### Design

For those users who want to explore deploy effect on some unforgeable name, **the user need to find
which block the deploy is packed and then explore the unforgeable name like `ExploreDeploy`**.

Send a request with fields below: 
```scala
final case class(
  blockHash: String,
  par: Par,
  usePreStateHash: Boolean
)
```
Response should be the same as for exploratory deploy `(Seq[Par], LightBlockInfo)`.

### Steps to implement the solution

- [x] Create new method on BlockAPI `getDataAtName` with corresponding tests
  #3544
- [x] Create gRPC binding for the new method
  #3592
- [x] Create HTTP binding for the new method
  #3584

### PR DESCRIPTION
**Pull request by stanislavlyalin**
## Overview

Added support for method `getDataAtPar` from `gRPC`. This method allows getting deploy results more effectively than `getDataAtName` because doesn't use the `depth` parameter.

Sample output (from pyrchain client):
```
payload {
  par {
    exprs {
      g_string: "data-at-par"
    }
  }
  block {
    blockHash: "70dedde39127f3bb9c4c0bbd8b51b6bf7235d170ee052f4bb2a6d4ad27f007ee"
    sender: "04415d35d799325f7784d7fbae15a3c4d0a24ff97429c7512b877b34bf3b7c801d220c38877e9a8a0736d168000ac5139cf505d9c030d1b7714dce4128ca3771b3"
    seqNum: 25
    sig: "3044022074f0ff8a4467bbc596166baa0283da5d75ab6fc4d1ebb149e0a0bee8ad686f3902201f90275bbaad719c6af6379e5e30b188e811a7ae33d0191ea932b8f4866b3677"
    sigAlgorithm: "secp256k1"
    shardId: "root"
    version: 1
    timestamp: 1643799057630
    parentsHashList: "5106c5a44b5b5a9888f02a3b9d90ac85afa0dd7aaa73c24abb9f58fd6dce848f"
    blockNumber: 25
    preStateHash: "00b4d835e6e37978cacc368a3f1ff8adaeb47d5f56d2c2e69e056b70a42cff94"
    postStateHash: "1651c6a64778a005d2f2188ca85b4ab91de46a7e75552c7800e9093e885432a8"
    bonds {
      validator: "04415d35d799325f7784d7fbae15a3c4d0a24ff97429c7512b877b34bf3b7c801d220c38877e9a8a0736d168000ac5139cf505d9c030d1b7714dce4128ca3771b3"
      stake: 150500000000000
    }
    blockSize: "127859"
    deployCount: 1
    faultTolerance: 1.0
    justifications {
      validator: "04415d35d799325f7784d7fbae15a3c4d0a24ff97429c7512b877b34bf3b7c801d220c38877e9a8a0736d168000ac5139cf505d9c030d1b7714dce4128ca3771b3"
      latestBlockHash: "5106c5a44b5b5a9888f02a3b9d90ac85afa0dd7aaa73c24abb9f58fd6dce848f"
    }
  }
}
```

### Notes

* Deploy should be done with such Rholang code
```
new return(`rho:rchain:deployId`) in { return!("data-at-par") }
```

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala [hunk 1]
```diff
@@ -803,8 +803,7 @@ object BlockAPI {
 
   def getDataAtPar[F[_]: Concurrent: EngineCell: Log: SafetyOracle: BlockStore](
       par: Par,
-      blockHash: String,
-      usePreStateHash: Boolean
+      blockHash: String
```
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala]: Parameter `usePreStateHash` should not be removed. It gives ability to query data on pre-state of the block.

File: casper/src/main/scala/coop/rchain/casper/util/comm/ListenAtPar.scala [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+  def toPar(name: Name): Par = Par(unforgeables = Seq(GUnforgeable(unforgToUnforgProto(name))))
+
+  private def unforgToUnforgProto(name: Name): GUnforgeable.UnfInstance = name match {
+    case UnforgPrivate(data)  => GPrivateBody(GPrivate(data.unsafeHexToByteString))
+    case UnforgDeploy(data)   => GDeployIdBody(GDeployId(data.unsafeHexToByteString))
+    case UnforgDeployer(data) => GDeployerIdBody(GDeployerId(data.unsafeHexToByteString))
+  }
+}
```
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/util/comm/ListenAtPar.scala]: Please split CLI part as a separate PR that we can finish later. We don't want to add additional command that doesn't work from the start.

File: models/src/main/protobuf/DeployServiceV1.proto [hunk 3]
```diff
@@ -133,6 +136,19 @@ message ListeningNameDataPayload {
   int32 length                         = 2;
 }
 
+// listenForDataAtPar
+message ListeningParDataResponse {
```
Comment (tgrospic) [inline models/src/main/protobuf/DeployServiceV1.proto]: We can call it like in WebAPI `RhoDataResponse` and correspondingly the payload also.

File: node/src/main/scala/coop/rchain/node/configuration/commandline/Options.scala [hunk 4]
```diff
@@ -741,6 +742,56 @@ final case class Options(arguments: Seq[String]) extends ScallopConf(arguments)
...
+    val content = opt[String](
+      required = true,
+      descr = "data for the provided type",
+      name = "content",
+      short = 'c'
+    )
+
+    // TODO: To match the code, you should use the solution from here
+    //  https://stackoverflow.com/questions/12078366/can-i-get-a-compile-time-list-of-all-of-the-case-objects-which-derive-from-a-sea
+    val allowedTypes = "UnforgPrivate, UnforgDeploy, UnforgDeployer"
```
Comment (tgrospic) [inline node/src/main/scala/coop/rchain/node/configuration/commandline/Options.scala]: If you want to find names of classes extending a trait you can use reflection.
```scala
import scala.reflect.runtime.universe

sealed trait Name
final case class UnforgPrivate(data: String)  extends Name
final case class UnforgDeploy(data: String)   extends Name
final case class UnforgDeployer(data: String) extends Name

universe.typeOf[Name].typeSymbol.asClass.knownDirectSubclasses.map(_.name.decodedName.toString)
// res0: scala.collection.immutable.Set[String] = Set(UnforgDeploy, UnforgDeployer, UnforgPrivate)
```

File: models/src/main/protobuf/DeployServiceV1.proto [hunk 5]
```diff
@@ -43,6 +44,8 @@ service DeployService {
   rpc getBlocks(BlocksQuery) returns (stream BlockInfoResponse) {}
   // Find data sent to a name.
   rpc listenForDataAtName(DataAtNameQuery) returns (ListeningNameDataResponse) {}
+  // Find data sent to a par.
+  rpc getDataAtPar(DataAtParQuery) returns (RhoDataResponse) {}
```
Comment (tgrospic) [inline models/src/main/protobuf/DeployServiceV1.proto]: Please rename this method to `getDataAtName` and add comment to _listenForDataAtName_:
_OBSOLETE: Use getDataAtName instead. This method will be removed in the future version of RNode._

File: models/src/main/protobuf/DeployServiceCommon.proto [hunk 6]
```diff
@@ -49,6 +49,12 @@ message DataAtNameQuery {
   Par name = 2;
 }
 
+message DataAtParQuery {
```
Comment (tgrospic) [inline models/src/main/protobuf/DeployServiceCommon.proto]: Please rename it to `DataAtNameByBlockQuery`.

File: models/src/main/protobuf/DeployServiceV1.proto (hunk 5 above)
Comment (stanislavlyalin) [reply models/src/main/protobuf/DeployServiceV1.proto]: @tgrospic, is comment `Find data sent to a par.` need to be updated to `Find data sent to a name.`?

Comment (stanislavlyalin) [reply models/src/main/protobuf/DeployServiceV1.proto]: Resolved in [commit](https://github.com/rchain/rchain/pull/3592/commits/274b51d06deab5b60643622dfe9438726fee3b36)

File: models/src/main/protobuf/DeployServiceCommon.proto (hunk 6 above)
Comment (stanislavlyalin) [reply models/src/main/protobuf/DeployServiceCommon.proto]: Resolved in [commit](https://github.com/rchain/rchain/pull/3592/commits/cbf57f7cf82cfa51dbb31c1fe1513d0821ad2bcd)

**Comment by stanislavlyalin** [comment]
bors try

File: models/src/main/protobuf/DeployServiceV1.proto (hunk 5 above)
Comment (tgrospic) [reply models/src/main/protobuf/DeployServiceV1.proto]: Yes, makes sense.

File: models/src/main/protobuf/DeployServiceCommon.proto [hunk 7]
```diff
@@ -49,6 +49,12 @@ message DataAtNameQuery {
   Par name = 2;
 }
 
+message DataAtNameByBlockQuery {
+  Par par              = 1 [(scalapb.field).no_box = true];
```
Comment (tgrospic) [inline models/src/main/protobuf/DeployServiceCommon.proto:53]: Kudos for no boxing annotation!

**Review by tgrospic — APPROVED** [review]
LGTM

Thanks for addressing all the comments.

**Review by zsluedem — APPROVED** [review]
LGTM

**Comment by stanislavlyalin** [comment]
bors merge

**Pull request merged**